homeassistant/components/todo/__init__.py

```diff
@@ -0,0 +1,310 @@
+        return sum([item.status == TodoItemStatus.NEEDS_ACTION for item in items])
+
+    @property
+    @abstractmethod
+    def todo_items(self) -> list[TodoItem] | None:
+        """Return the To-do items in the To-do list."""
+
+    async def async_create_todo_item(self, item: TodoItem) -> None:
+        """Add an item to the To-do list."""
+        raise NotImplementedError()
+
+    async def async_update_todo_item(self, item: TodoItem) -> None:
+        """Add an item to the To-do list."""
+        raise NotImplementedError()
+
+    async def async_delete_todo_items(self, uids: set[str]) -> None:
+        """Add an item to the To-do list."""
+        raise NotImplementedError()
+
+    async def async_move_todo_item(self, uid: str, previous: str | None) -> None:
+        """Move an item in the To-do list."""
+        raise NotImplementedError()
+
+
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "todo/item/list",
+        vol.Required("entity_id"): cv.entity_id,
+    }
+)
+@websocket_api.async_response
+async def websocket_handle_todo_item_list(
+    hass: HomeAssistant, connection: websocket_api.ActiveConnection, msg: dict[str, Any]
+) -> None:
+    """Handle the list of To-do items in a To-do- list."""
+    component: EntityComponent[TodoListEntity] = hass.data[DOMAIN]
+    if (
+        not (entity_id := msg.get(CONF_ENTITY_ID))
+        or not (entity := component.get_entity(entity_id))
+        or not isinstance(entity, TodoListEntity)
+    ):
+        connection.send_error(msg["id"], ERR_NOT_FOUND, "Entity not found")
+        return
+
+    items: list[TodoItem] = entity.todo_items or []
+    connection.send_message(
+        websocket_api.result_message(
+            msg["id"], {"items": [dataclasses.asdict(item) for item in items]}
+        )
+    )
+
+
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "todo/item/create",
+        vol.Required("entity_id"): cv.entity_id,
+        vol.Required("item"): WEBSOCKET_TODO_ITEM_CREATE_SCHEMA,
+    }
+)
+@websocket_api.async_response
+async def websocket_handle_todo_item_create(
+    hass: HomeAssistant, connection: websocket_api.ActiveConnection, msg: dict[str, Any]
+) -> None:
+    """Handle creation of a To-do item on a To-do list."""
+    component: EntityComponent[TodoListEntity] = hass.data[DOMAIN]
+    if not (entity := component.get_entity(msg["entity_id"])):
+        connection.send_error(msg["id"], ERR_NOT_FOUND, "Entity not found")
+        return
+
+    if (
+        not entity.supported_features
+        or not entity.supported_features & TodoListEntityFeature.CREATE_TODO_ITEM
+    ):
+        connection.send_message(
+            websocket_api.error_message(
+                msg["id"],
+                ERR_NOT_SUPPORTED,
+                "To-do list does not support To-do item creation",
+            )
+        )
+        return
+
+    try:
+        await entity.async_create_todo_item(item=TodoItem(**msg["item"]))
```

Comment:
We pin the frontend version to the core version but maybe still nice to be explicit and name the keyword arguments?